### PR TITLE
[Multi-Tab] Detect and recover from GCed Remote Document Changelog

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This is meant to address the case when one tab removed document changes from the document change log that a throttled tab has not yet seen. The detection relies on the fact that change IDs are incremented by 1 each time (as guaranteed by https://www.w3.org/TR/IndexedDB/#key-generator-construct)

### CODE REVIEW

File: packages/firestore/src/local/indexeddb_remote_document_cache.ts [hunk 1]
```diff
@@ -172,18 +167,26 @@ export class IndexedDbRemoteDocumentCache implements RemoteDocumentCache {
...
+          // client may have garbage collected the remote document changelog.
+          if (this._lastProcessedDocumentChangeId + 1 !== documentChange.id) {
+            return PersistencePromise.reject(
+              new FirestoreError(
+                Code.DATA_LOSS,
+                REMOTE_DOCUMENT_CHANGELOG_GARBAGE_COLLECTED_ERR_MSG
+              )
+            );
+          }
+          firstIteration = false;
```
Comment (mikelehen) [inline packages/firestore/src/local/indexeddb_remote_document_cache.ts]: Per my comment elsewhere, perhaps rework this so that getNewDocumentChanges() always advances past all document changes, even if it ultimately returns an error...  e.g. by changing firstIteration to minChangeId and checking it in the next() block...

File: packages/firestore/src/local/indexeddb_remote_document_cache.ts [hunk 2]
```diff
@@ -217,6 +220,40 @@ export class IndexedDbRemoteDocumentCache implements RemoteDocumentCache {
...
+      { keysOnly: true, reverse: true },
+      (key, value, control) => {
+        this._lastProcessedDocumentChangeId = key;
+        control.done();
+      }
+    );
+  }
+}
+
+export function isRemoteDocumentChangesGarbageCollectedError(
```
Comment (mikelehen) [inline packages/firestore/src/local/indexeddb_remote_document_cache.ts]: I like verbosity, but "is...Error" gets hard to parse when there are 5 words in between.  How about isDocumentChangeMissingError()?

I was also wondering if you could just:
```
export const DOCUMENT_CHANGE_MISSING_ERROR = new Error(...);
```

But I am not sure how JS works. Would that mess up the stack trace in the error or something?

It's also a little dubious to be using errors for control flow.  We could change getNewDocumentChanges() to return null if it encountered a missing change or something, but I think I'm not that thrilled about the idea.

File: packages/firestore/test/unit/local/test_remote_document_cache.ts [hunk 3]
```diff
@@ -67,13 +68,42 @@ export class TestRemoteDocumentCache {
...
-      'getNextDocumentChanges',
+      'getNewDocumentChanges',
       'readonly',
       txn => {
         return this.cache.getNewDocumentChanges(txn);
       }
     );
   }
+
+  resetLastProcessDocumentChange(): Promise<void> {
```
Comment (mikelehen) [inline packages/firestore/test/unit/local/test_remote_document_cache.ts]: I'm hoping this method goes away, but if not you're missing 'ed' here and below.

File: packages/firestore/test/unit/local/test_remote_document_cache.ts [hunk 4]
```diff
@@ -67,13 +68,42 @@ export class TestRemoteDocumentCache {
...
+      throw new Error(
+        'Can only removeDocumentChangesThroughChangeId() in IndexedDb'
+      );
+    }
+    return this.persistence.runTransaction(
+      'removeDocumentChangesThroughChangeId',
+      'readwrite-primary',
+      txn => {
+        return (this
+          .cache as IndexedDbRemoteDocumentCache).removeDocumentChangesThroughChangeId(
```
Comment (mikelehen) [inline packages/firestore/test/unit/local/test_remote_document_cache.ts]: Is this cast necessary?  If you moved your if check above into this closure it probably wouldn't be...

File: packages/firestore/src/local/indexeddb_remote_document_cache.ts (hunk 1 above)
Comment (schmidt-sebastian) [reply packages/firestore/src/local/indexeddb_remote_document_cache.ts]: Consider it done.

File: packages/firestore/src/local/remote_document_cache.ts [hunk 5]
```diff
@@ -94,4 +94,15 @@ export interface RemoteDocumentCache {
...
   ): PersistencePromise<MaybeDocumentMap>;
+
+  /**
+   * Skips all existing change log entries in IndexedDb and moves the change log
+   * cursor past the last existing change. This method should only be called if
+   * `getNewDocumentChanges()`can no longer replay all changes and all views
+   * have already been manually synchronized.
+   */
+  // PORTING NOTE: This is only used for multi-tab synchronization.
+  resetLastProcessedDocumentChange(
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/local/remote_document_cache.ts]: I removed the method and inlined the call that resets the change log pointer.

File: packages/firestore/test/unit/local/test_remote_document_cache.ts (hunk 3 above)
Comment (schmidt-sebastian) [reply packages/firestore/test/unit/local/test_remote_document_cache.ts]: And it's gone!

File: packages/firestore/test/unit/local/test_remote_document_cache.ts (hunk 4 above)
Comment (schmidt-sebastian) [reply packages/firestore/test/unit/local/test_remote_document_cache.ts]: No, I managed to get rid of it by moving the assert.

File: packages/firestore/src/local/indexeddb_remote_document_cache.ts (hunk 2 above)
Comment (schmidt-sebastian) [reply packages/firestore/src/local/indexeddb_remote_document_cache.ts]: I changed the name to `isDocumentChangeMissingError`.

In general, I would like to shy away from statically allocating errors as it does mess up the stacktraces (even in JS).

I'm gonna use the current fire alarm in SFO as an excuse to not comment on the API contract. I can't really think of a much better way to convey this information.

File: packages/firestore/src/local/indexeddb_remote_document_cache.ts [hunk 6]
```diff
@@ -171,16 +171,22 @@ export class IndexedDbRemoteDocumentCache implements RemoteDocumentCache {
...
           // client may have garbage collected the remote document changelog.
           if (this._lastProcessedDocumentChangeId + 1 !== documentChange.id) {
-            return PersistencePromise.reject(
-              new FirestoreError(
-                Code.DATA_LOSS,
-                REMOTE_DOCUMENT_CHANGELOG_GARBAGE_COLLECTED_ERR_MSG
+            // Reset the `lastProcessedDocumentChangeId` to allow further
+            // invocations to successfully return the changes after this
+            // rejection.
+            return this.synchronizeLastDocumentChangeId(changesStore).next(() =>
```
Comment (mikelehen) [inline packages/firestore/src/local/indexeddb_remote_document_cache.ts]: Just to verify, did you consider just letting this iteration continue so that line 199 can correctly update _lastProcessedDocumentChangeId, and then we could (I think?) get rid of the synchronizeLastDocumentChangeId?   I think the reduced code complexity would outweigh any potential performance saving of getting to skip the changedKeys accounting...

**Comment by schmidt-sebastian** [comment]
I tested this PR manually this morning. I found one area of improvement (we should set firstIteration to false before we return, since the iterate function just keeps iterating).

Outside of this PR, I did find a slew of things that were not working/no longer working in Multi-Tab, so thanks for asking for an extra QA run :)
